lib/ListResource.js

```diff
@@ -0,0 +1,71 @@
+var merge           = require('./utils/merge');
+var parseLinkHeader = require('./utils/parse_link_header');
+var errors          = require('./Errors');
+var Thenable        = require('./vendor/Thenable');
+
+function ListResource(object, path, opts, client) {
+  this.object = object;
+  this.path = path;
+  this.opts = opts;
+  this.client = client;
+  this.interation_count = 0;
+  this.done = false;
+}
+
+merge(ListResource.prototype, {
+  autoPagingEach: function(onItem, opts) {
+    return new Thenable(function(resolve, reject) {
+      this.auto_handle_rate_limits = ( !opts || opts.auto_handle_rate_limits === undefined ? true : opts.auto_handle_rate_limits );
+      this.auto_handle_rate_limits_callback = ( !opts || opts.auto_handle_rate_limits_callback === undefined ? true : opts.auto_handle_rate_limits_callback );
+      if (this.done) {
+        this._onError(errors('pagination', 'pagination completed'));
+      }
+
+      this._onDone = function() {
+        resolve();
+      };
+
+      this._onError = function(error) {
+        if (this.auto_handle_rate_limits) {
+          setTimeout(this._getResponse.bind(this), error.retryAfter * 1000);
+          this.auto_handle_rate_limits_callback(error.retryAfter);
```

Comment:
Thoughts on including a max number of retry attempts before giving up? @thierryung

Reply:
Yeah, could be useful there. Adding.